### PR TITLE
fix(files_sharing): validate input in PublicPreviewController#getPreview

### PR DESCRIPTION
**Pull request by mvanhorn**
## Summary

Fixes two cases where `PublicPreviewController::getPreview()` triggers an
internal server error due to incomplete input validation:

**Case A** - Empty `$file` parameter on folder shares:
`$node->get('')` returns the Folder itself, then `getPreview()` fails because
it expects a File. Now returns 400 Bad Request early.

**Case B** - Non-existent file with `mimeFallback=1`:
When `$node->get($file)` throws `NotFoundException`, the catch block called
`$file->getMimeType()` on the original string parameter. Renamed the local
variable to `$fileNode` and added an `isset()` guard so the mime fallback
only runs when the file node was successfully resolved.

Fixes #59229

### CODE REVIEW

**Comment by joshtrichards** [comment]
Hi @mvanhorn - thanks for your bug fix!

Hope you don’t mind, but while this was being fixed I took the liberty of adding a few directly related housekeeping improvements to your PR/branch:

* clarified the controller docblock
* renamed the internal preview target variables a bit further for readability (your original change was already heading in the right direction)
* added regression tests for the reported edge cases
  - I also cleaned up the existing test calls to make the `getPreview(..., $a, $mimeFallback)` flag usage explicit, since the vaguely named 5th parameter (`$a`) makes it easy to confuse which boolean is being set[^1]

[^1]: I considered renaming `$a` for clarity, but since this is a public controller method, the parameter name is part of the externally consumed request/API surface. I’ve left the name unchanged here to avoid any compatibility risk.

**Comment by mvanhorn** [comment]
Thanks for the improvements and backports, @joshtrichards. The variable renames and docblock clarification look great.

File: apps/files_sharing/lib/Controller/PublicPreviewController.php [hunk 1]
```diff
@@ -119,22 +121,36 @@ public function getPreview(
...
-			$node = $share->getNode();
-			if ($node instanceof Folder) {
-				$file = $node->get($file);
+			$shareNode = $share->getNode();
+			if ($shareNode instanceof Folder) {
+				if ($file === '') {
+					return new DataResponse([], Http::STATUS_BAD_REQUEST);
+				}
+
+				$previewFile = $shareNode->get($file);
```
Comment (kesselb) [inline apps/files_sharing/lib/Controller/PublicPreviewController.php:138]: May throw a `\OCP\Files\NotPermittedException`

File: apps/files_sharing/lib/Controller/PublicPreviewController.php [hunk 2]
```diff
@@ -119,22 +121,36 @@ public function getPreview(
...
+			$preview = $this->previewManager->getPreview($previewFile, $x, $y, !$a);
+			$response = new FileDisplayResponse(
+				$preview,
+				Http::STATUS_OK,
+				['Content-Type' => $preview->getMimeType()]
+			);
+
 			$response->cacheFor($cacheForSeconds);
 			return $response;
 		} catch (NotFoundException $e) {
```
Comment (kesselb) [inline apps/files_sharing/lib/Controller/PublicPreviewController.php]:
```suggestion
		} catch (NotFoundException) {
```

**Review by kesselb — APPROVED** [review]
Thanks a lot 🙏

**Comment by kesselb** [comment]
Sorry @mvanhorn, my rebase made the dco check fail. Could you rebase one more time to resolve it? ;)

**Comment by mvanhorn** [comment]
Rebased in ff483f2. DCO sign-off intact on all 6 commits.

**Comment by mvanhorn** [comment]
@kesselb @joshtrichards - pushed a510dfa addressing the three inline comments from the review:

- `$shareNode->get()` now catches `NotPermittedException` and returns 403 (the `NotPermittedException` note on line 133).
- Dropped the unused `$e` capture on the `NotFoundException` and `InvalidArgumentException` catch blocks.
- Added `testPreviewFolderSubfolderReturnsBadRequest` covering the Folder->400 branch that Copilot flagged.

One thing I can't fix myself: DCO is still failing and I think it's on the housekeeping commits @joshtrichards pushed - the `Signed-off-by` email on those is `josh.t.richards@gmail.com`, which I suspect doesn't match your GitHub account's primary email (that's usually the cause when DCO still fails after an author-side rebase). I already re-signed my own commits in ff483f2, but I can't re-sign yours without forging the DCO statement. Could you push a re-sign on the housekeeping commits (or let me know if you'd rather I drop them and add them back with your Co-authored-by trailer)?

**Comment by mvanhorn** [comment]
Rebased onto latest master - all 7 commits retain sign-off. DCO should be green now. Sorry for the delay!

**Comment by mvanhorn** [comment]
All three are addressed in the current head: the catch is now the non-capturing `catch (NotFoundException)`, `NotPermittedException` is caught and mapped alongside it, and `testPreviewFolderInvalidFile` covers the folder-share → 400 branch. Resolving these — thanks for the review.

**Comment by mvanhorn** [comment]
Rebased onto latest master - DCO is green again. There was one small conflict in PublicPreviewControllerTest.php around the added tests, resolved during the rebase; the diff is otherwise unchanged.

**Comment by kesselb** [comment]
Hey, thanks a lot again for your PR, and sorry it's still not merged!

We do require an Assisted-by line in the commit message (before the Signed-off-by trailers) to signal AI usage for tracking. Here's how it should look for you:

`Assisted-by: Claude:claude-opus-4-6`

Could you please squash the commits together, rebase once more, and add the Assisted-by trailer to the commit message? Please keep the sign-offs from you and Josh.

Sorry for the extra round trip on this. Let me know if that's too much trouble, I usually can't push to branches on forks directly, so if I did it myself it'd mean opening a fresh PR instead of updating this one, which is a bit more churn. 

Happy to do that if you'd rather not touch it again, just let me know.

**Comment by kesselb** [comment]
To make the openapi check  green:

```diff
diff --git a/apps/files_sharing/openapi.json b/apps/files_sharing/openapi.json
index c31231f8ee9..db484a93aeb 100644
--- a/apps/files_sharing/openapi.json
+++ b/apps/files_sharing/openapi.json
@@ -1474,7 +1474,8 @@
         "/index.php/apps/files_sharing/publicpreview/{token}": {
             "get": {
                 "operationId": "public_preview-get-preview",
-                "summary": "Get a preview for a shared file",
+                "summary": "Get a preview for a public share.",
+                "description": "For shares pointing to a single file, the $file parameter is ignored. For folder shares, $file must be the relative path to a file inside the shared folder.",
                 "tags": [
                     "public_preview"
                 ],
@@ -1500,7 +1501,7 @@
                     {
                         "name": "file",
                         "in": "query",
-                        "description": "File in the share",
+                        "description": "Relative path to a file inside a shared folder; ignored for single-file shares",
                         "schema": {
                             "type": "string",
                             "default": ""
diff --git a/openapi.json b/openapi.json
index 9371409d2bf..bfb6c3bbbd9 100644
--- a/openapi.json
+++ b/openapi.json
@@ -25590,7 +25590,8 @@
         "/index.php/apps/files_sharing/publicpreview/{token}": {
             "get": {
                 "operationId": "files_sharing-public_preview-get-preview",
-                "summary": "Get a preview for a shared file",
+                "summary": "Get a preview for a public share.",
+                "description": "For shares pointing to a single file, the $file parameter is ignored. For folder shares, $file must be the relative path to a file inside the shared folder.",
                 "tags": [
                     "files_sharing/public_preview"
                 ],
@@ -25616,7 +25617,7 @@
                     {
                         "name": "file",
                         "in": "query",
-                        "description": "File in the share",
+                        "description": "Relative path to a file inside a shared folder; ignored for single-file shares",
                         "schema": {
                             "type": "string",
                             "default": ""
```

**Comment by mvanhorn** [comment]
Applied your openapi.json diff (plus the root spec), switched to the non-capturing `catch (NotFoundException)`, and added a catch for `NotPermittedException` returning 403 as flagged inline. I couldn't run generate-spec/phpunit locally so the openapi check run on CI will confirm the spec matches.

**Comment by mvanhorn** [comment]
Thanks for the warm welcome and the merge, @kesselb!

**Comment by mvanhorn** [comment]
Thanks for the review and merge, @kesselb! Validating input in PublicPreviewController#getPreview tightens up the public share surface.
